### PR TITLE
Machine ID CA rotation

### LINKED ISSUE #10891: Add CA rotation support
CA rotation support is missing from Machine ID.

### PR DESCRIPTION
**Pull request by strideynet**
Closes #10891 

Updates `tbot` to handle CA rotations.

## Implementation

My gut feeling is that triggering the standard renewal mechanism is the best course of action here. It avoids having distinct logic for fetching the new CAs and certificates issued by them, and prevents the risk of a normal renewal run conflicting with an in-progress CA rotation.

The auth server holds the logic as to what CAs and certificates are provisioned. As long as tbot fetches the CAs and certificates as it would normally in a renewal but during a CA rotation stage, it will receive copies of the old and new CA where appropriate. From testing, I can see that both these certificates are concatenated in the files output by tbot (e.g `teleport-host-ca.crt`)

## Other

Do we want to backport this?

## Todo

- [ ] Ensure that various clients accept the CA certificate files where both old and new CA are concatenated.
- [x] Fix `known_hosts` generating with both entries on the same line (when there are two trusted CA certificates).

### CODE REVIEW

**Comment by zmb3** [comment]
> Do we want to backport this?

Not necessary for v9, as Machine ID was only in preview. If this doesn't get merged before we create `branch/v10` then yes, we'll want to backport for v10.

File: lib/tbot/ca_rotation.go
Comment (strideynet) [inline lib/tbot/ca_rotation.go:60]: As far as I can tell it's correct to filter by the cluster name on all of these, but I need someone more familiar than me with our CA to confirm this.

File: lib/tbot/ca_rotation.go
Comment (strideynet) [inline lib/tbot/ca_rotation.go:135]: Checking the cluster name and the CA type here feels redundant as the Watcher should only send us the events we ask for, can I safely remove these checks ?? Or is there known cases where the Watcher does not filter as expected ?

File: lib/tbot/ca_rotation.go
Comment (strideynet) [reply lib/tbot/ca_rotation.go:107]: That'll work nicely, thanks.

**Comment by strideynet** [comment]
Just as a quick update to where we are we this PR. As it stands, CA rotation now works (when testing on my local machine). There's a few outstanding items...

Firstly, I need to work on the retry looping for the entire CA watcher. I want to have something that will detect X number of failures in the last Y minutes, and exit out if this is the case. I want to ensure that the CA watcher isn't failing silently in the background, and that users will be sufficiently aware if something is misconfigured and causing the watcher to fail repeatedly. It would be very painful if this first manifested itself when a customer went to complete a rotation.

The second part is I need to come up with a nicer way of debouncing multiple CA rotations. In some circumstances, all four CAs will be rotated at once, and this causes multiple events to be received. We should ensure this doesn't cause multiple superfluous renewals. Another issue, is that some CA renewal transitions cause the server to reload and drop connections. As far as I can tell, the events related to the renewal are sent /before/ this reload occurs. This means in the right conditions that the renewal could begin as the server goes away, and this would cause `tbot` to crash.

Therefore, my current feeling is that the solution may be to trigger the renewal some X seconds (potentially 10-15 seconds) after the events are received. Further events received between the first event triggering the countdown, and the renewal occurring, would be ignored. This solves two things:
- It reduces the likelihood that the renewal will occur during the server reload caused by the rotation
- It debounces the multiple events.

It feels like we would also need to add some retry logic to the renewal loop, allowing a failed renewal to be tried some X times before crashing the program out (as it currently does)

File: lib/tbot/ca_rotation.go
Comment (strideynet) [inline lib/tbot/ca_rotation.go:44]: I've gone for a Timer/Mutex based debouncer here. I originally wrote a channel based implementation:

```go
    debounceChan := make(chan *struct{})
    defer close(debounceChan)
    go func() {
        for {
            select {
            case v := <-debounceChan:
                if v == nil {
                    return
                }
                t := time.NewTimer(time.Second * 10)
                for {
                    select {
                    case <-ctx.Done():
                        return
                    case <-t.C:
                        select {
                        case b.reloadChan <- struct{}{}:
                        default:
                            b.log.Debugf("Renewal already queued, ignoring reload request.")
                        }
                    case v := <-debounceChan:
                        if v == nil {
                            return
                        }
                        // swallow other reload msgs whilst we are in the debounce
                        // period.
                    }
                }
            }
        }
    }()
```

But the channel based implementation seemed a bit more verbose and less clear about what it actually did, even if it handles the context cancellation sooner.

File: lib/tbot/ca_rotation.go
Comment (strideynet) [inline lib/tbot/ca_rotation.go:87]: TODO: Establish if ten seconds is an appropriate debounce period for most deployments.

**Comment by strideynet** [comment]
I've added retries with limits for the CA rotation watching and the main renewal loop. However, I'm a little uncertain if this is behaviour we desire. Previously, I've tended to implement things that will retry some sane amount before then crashing out and letting the supervisor (systemd, kubernetes, whatever has been invented this week) decide how to handle the situation. 

Looking across the Teleport codebase, it feels like we currently prefer to continue retrying within the process indefinitely. Can anyone confirm or deny what the preferred behaviour is here? I think both sides have merits, but it would be nice to pick one that maintains consistency.

**Comment by strideynet** [comment]
**Status Update**

This PR works, bar the test. My current issue is it looks like the in-memory test auth server does not properly reload when a CA rotation occurs. This means the bot fails to create a new client, as when it connects after entering the `standby` phase, the certificates presented by the server are still signed by the old CA, which is no longer trusted by the bot.

**Comment by strideynet** [comment]
> **Status Update**
> 
> This PR works, bar the test. My current issue is it looks like the in-memory test auth server does not properly reload when a CA rotation occurs. This means the bot fails to create a new client, as when it connects after entering the `standby` phase, the certificates presented by the server are still signed by the old CA, which is no longer trusted by the bot.

I've managed to solve this by using `service.Run` which invokes the mechanism that handles in-process reloads. The test now works locally, just seeing if it works in CI.

File: lib/tbot/ca_rotation.go
Comment (timothyb89) [inline lib/tbot/ca_rotation.go:111]: Hmm. I may just not fully understand the flow here, but I wonder if this is really something that should have a retry limit. Would it be terribly inappropriate to keep trying forever (with sensible backoffs)?

As I understand the outer loop we'll only currently refresh the attempt counter when the watcher finishes on its own. If a bot is running for a long time (I'd argue weeks/months of uptime is reasonable) it seems like we might eventually stop handling CA rotations due to accumulated errors even if we were happily listening for events 99% of the time.

Alternatively, maybe we could reset the event counter on some condition, or if we do want it to have a limit, it might be good to pretty loudly log that we'll no longer respond to CA events for the remainder of the session.

**Review by timothyb89 — APPROVED** [review]
Looks really solid overall!

**Comment by strideynet** [comment]
> This seems reasonable. I don't think there are too many situations where we could really get caught in between rotations, and I think we'll just renew again after 10 seconds at worst?

Yeah so my intention with a lot of this was to avoid scenarios where we missed a rotation, at the cost of potentially renewing once or twice too much. Whilst this isn't perfect from a performance point of view, it's certainly preferable to losing a `tbot` off of the cluster, especially as user's don't have insight into how many `tbot`s are currently connected.

File: lib/tbot/ca_rotation.go
Comment (strideynet) [reply lib/tbot/ca_rotation.go:111]: From what I can recall, my original intention with this behaviour was to crash out after enough failed attempts, and defer to the superviser (systemd etc). However, looking at the code, that's not what I've implemented here at all. I think with the current implementation, it will go through a series of back offs, before resetting the back off and trying again with the backoff back to a minimal amount, which doesn't seem right. I'll rectify this.

**Comment by marcoandredinis** [comment]
@strideynet A couple of weeks ago we removed some CA Rotation tests that were part of the CA Signing Algorithm
If you need inspiration, you can take a look here https://github.com/gravitational/teleport/pull/13033/files
Search for `testRotateChangeSigningAlg`

**Comment by strideynet** [comment]
We should hold off on merging this down until v10.0.0 has been cut and released, as we'll want this to go out in 10.1

**Comment by strideynet** [comment]
Gently prodding @gabrielcorado and @zmb3 for a review, we should be unblocked on merging this now that 10.0.0 is cut and this PR is starting to block other work.

File: lib/tbot/ca_rotation.go
Comment (zmb3) [reply lib/tbot/ca_rotation.go:200]: @rosstimothy can you answer this one?

File: lib/tbot/tbot_test.go
Comment (zmb3) [inline lib/tbot/tbot_test.go:137]: I don't feel super strongly, so feel free to push back, but I generally use a wait group when there's a group (ie more than one) thing I want to wait for.

In this case, a simple channel read would also work, no?

File: lib/tbot/tbot_test.go
Comment (zmb3) [inline lib/tbot/tbot_test.go:160]: How do we know that all of these sleeps will be long enough to be reliable and not flaky?

File: lib/tbot/tbot_test.go
Comment (strideynet) [reply lib/tbot/tbot_test.go:137]: There is actually multiple things in the group :') It's passed into `setupServerForCARotationTest` but I haven't really helped my case by obscuring that there. Essentially, it's a group of the bot and the running Teleport process, so that I can ensure both have closed down.

File: lib/tbot/tbot_test.go
Comment (strideynet) [reply lib/tbot/tbot_test.go:160]: Being frank, we don't. It's a bit rubbish, but I was worried about adding more complexity to this PR (in the form of some kind of event emitter) purely to service the tests.  Ten seconds is very generous here for the amount of things it needs to set up though.

Whilst a bit nebulous, it is my full intention to get around to adding this sort of "Ready" event emitter. One thing I have in mind for Machine ID is potentially a model where user's writing Go would be able to directly embed it within their codebase, and this sort of ready event would be needed by them.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**
